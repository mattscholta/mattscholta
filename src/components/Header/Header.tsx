import * as React from 'react';
import { color } from '@barguide/style-guide';
import classnames from 'classnames';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { useRecoilState } from 'recoil';

import { Logo } from '@src/components/Logo';
import { data } from '@src/data/header';
import { Hamburger, Octocat } from '@src/icons';
import { appAtom } from '@src/recoil/atoms/app';
import { analytics } from '@src/utils/analytics';

import styles from './styles.module.css';

export interface HeaderProps {}

/**
 * @name Header
 * @description Primary Navigation Header
 */
export const Header = (_props: HeaderProps) => {
  // Hooks
  const [app, setApp] = useRecoilState(appAtom);
  const { navigation } = app;
  const { pathname } = useRouter();

  // Setup
  const isOpen = navigation.open;

  // Styles
  const cssNav = classnames(styles.nav, {
    [styles.open]: isOpen
  });

  // Markup
  const onToggleNav = () => {
    setApp({ ...app, navigation: { ...navigation, open: !isOpen } });
  };

  const onTrackEvent = (page: any) => () => {
    analytics.event('nav-header', 'click', page);
  };

  // Markup
  const renderLink = (node: any) => {
    const { children, href } = node;
    const key = `nav-primary-${children}`;

    const isHome = pathname === '/' && href === '/';
    const isOther = !isHome && href !== '/' && pathname.startsWith(href);

    const cssLink = classnames(styles.link, {
      [styles.active]: isHome || isOther
    });

    return (
      <Link href={href} key={key}>
        <a className={cssLink} onClick={onToggleNav}>
          {children}
        </a>
      </Link>
    );
  };

  return (
    <div className={styles.component}>
      <div className="ui-container-xl u-flex u-justify-center md:u-justify-start">
        <button
          className={styles.menu}
          onClick={onToggleNav}
          onKeyPress={onToggleNav}
          type="button"
        >
          <Hamburger fill={color.white} height={20} width={20} />
        </button>
        <Link href="/">
          <a className="md:u-mr-2x">
            <Logo className={styles.svgBrand} fill={color.white} />
          </a>
        </Link>
        <nav className={cssNav}>{data.map(renderLink)}</nav>
        <a
          href="https://www.github.com/visormatt"
          onClick={onTrackEvent('github')}
          onKeyPress={onTrackEvent('github')}
          rel="noopener noreferrer"
          target="_blank"
          title="Find me on Github"
        >
          <Octocat
            className={styles.svgOctocat}
            color={color.gray[8]}
            fill={color.white}
          />
        </a>
      </div>
    </div>
  );
};
