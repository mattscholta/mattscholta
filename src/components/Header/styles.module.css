.component :root {
  --spot-dropshadow: rgba(0 0 0 / 30%);
}

.component {
  background-color: var(--spot-grey-8);
  border-bottom: unset;
  box-shadow: 0 3px var(--grid-1xs) 0 var(--spot-dropshadow);
  height: var(--header-height);
  left: 0;
  min-width: var(--min-width);
  padding: 0 theme('spacing.2x');
  position: fixed;
  right: 0;
  top: 0;
  z-index: 9;
}

.nav {
  overflow: auto;
  white-space: nowrap;
}

.link {
  color: var(--spot-white);
  display: inline-block;
  font-size: theme('fontSize.14');
  line-height: var(--header-height);
  padding: 0 var(--grid-2x);
  position: relative;
  text-decoration: none;
  transition: color 140ms ease-out;
}

.link:after {
  background-color: var(--spot-grey-3);
  border-top-left-radius: 2px;
  border-top-right-radius: 2px;
  bottom: 0;
  content: '';
  height: 0;
  left: 0;
  position: absolute;
  right: 0;
  transition: height 140ms ease-out;
}

.link:hover:after,
.active:after {
  height: 3px;
}

.link:hover {
  color: var(--spot-white);
  text-decoration: none;
}

.active:after {
  background-color: var(--spot-white);
}

.logo,
.svgBrand,
.svgOctocat {
  display: inherit;
}

.logo {
  margin: 0 var(--grid-2x) 0 var(--grid-3x);
}

.menu {
  background: transparent;
  border: none;
  display: none;
}

.active,
.active:hover {
  color: var(--spot-white);
  text-decoration: none;
}

.svgBrand {
  height: 50px;
  width: 50px;
}

.svgOctocat {
  height: var(--header-height);
  position: fixed;
  right: 0;
}

@media (max-width: 768px) {
  .logo {
    margin: auto;
  }

  .menu {
    cursor: pointer;
    display: inline-flex;
    left: 0;
    padding: var(--grid-2xs);
    padding: 14px 12px;
    position: fixed;
    top: 0;
  }

  .nav {
    background-color: var(--spot-grey-8);
    bottom: 0;
    box-shadow: 0 3px var(--grid-1xs) 0 var(--spot-dropshadow);
    height: calc(100vh - var(--header-height));
    left: 0;
    opacity: 0;
    overflow: hidden;
    padding-top: 20px;
    pointer-events: none;
    position: absolute;
    right: 0;
    top: var(--header-height);
    transform: translate3d(0, 0, 0);
    transition: opacity 340ms ease-in-out;
    width: 100%;
  }

  .nav.open {
    opacity: 1;
    pointer-events: all;
  }

  .link {
    background-color: transparent;
    display: block;

    /* prettier-ignore */
    transition: background-color 200ms ease-out, color 200ms ease-out, padding-left 200ms ease-out;
  }

  .active {
    background-color: theme('colors.gray.7');
    color: theme('colors.gray.2');
    padding-left: theme('spacing.3x');
  }

  .link:hover:after,
  .active:after {
    height: 0;
  }
}
