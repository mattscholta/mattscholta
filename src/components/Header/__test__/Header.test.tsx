import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { RecoilRoot } from 'recoil';

import { Header, HeaderProps } from '../Header';

jest.mock('next/router', () => ({
  useRouter() {
    return {
      asPath: '',
      pathname: '',
      push: jest.fn(),
      query: {},
      route: '/'
    };
  }
}));

describe('Header', () => {
  const navigationToggle = jest.fn();

  let props: HeaderProps;
  let wrapper: RenderResult;

  const createWrapper = () => (
    <RecoilRoot>
      <Header {...props} />
    </RecoilRoot>
  );

  beforeEach(() => {
    props = {
      navigation: {
        open: false
      },
      navigationToggle
    };

    wrapper = render(createWrapper());
  });

  it('matches snapshot', () => {
    expect(wrapper.container.childNodes).toMatchSnapshot();
  });
});
