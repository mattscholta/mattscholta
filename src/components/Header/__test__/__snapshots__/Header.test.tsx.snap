// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
NodeList [
  <div
    class="component"
  >
    <div
      class="ui-container-xl u-flex u-justify-center md:u-justify-start"
    >
      <button
        class="menu"
        type="button"
      >
        <svg
          height="20"
          id="my_element"
          viewBox="0 0 20 20"
          width="20"
        >
          <title>
            Menu
          </title>
          <rect
            fill="#fff"
            height="2.5"
            width="20"
            x="0%"
            y="0%"
          />
          <rect
            fill="#fff"
            height="2.5"
            width="20"
            x="0%"
            y="8.75"
          />
          <rect
            fill="#fff"
            height="2.5"
            width="20"
            x="0%"
            y="17.5"
          />
        </svg>
      </button>
      <a
        class="md:u-mr-2x"
        href="/"
      >
        <svg
          class="svgBrand"
          height="59"
          viewBox="0 0 100 59"
          width="100"
        >
          <title>
            mattscholta
          </title>
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="none"
            stroke-width="1"
          >
            <path
              d="M72.573561,53.8100014 L67.7974434,43 L63.0213258,53.8100014 L57.7974434,53.8100014 L52.573561,53.8100014 L47.7974434,43 L43.0213258,53.8100014 L24.0228956,53.8100014 L47.7974434,0 L57.7974434,22.6334489 L67.7974434,0 L91.5719913,53.8100014 L72.573561,53.8100014 Z M7.10754499,21.7630003 L16.7229599, 8.8817842e-16 L19.5285978,0 L8.51036391,24.9380633 L7.10754499, 21.7630003 Z M9.82164089,28.3034712 L22.3267926,2.0539126e-15 L26.1647839,0 L11.7406366,32.6468202 L9.82164089,28.3034712 Z M28.7229599,0 L44.5719913,0 L20.7974434,53.8100014 L16.7974434, 44.7566218 L36.5719913, 0 L28.7229599,0 L28.7229599,8.8817842e-16 Z M28.7229599,8.8817842e-16 L28.7229599, 8.8817842e-16 L34.1647839, 0 L15.6106287,41.9944523 L12.8897168,35.83609 Z M5.35664415, 14.1239332 L0,2 L10.7132883,2 L5.35664415,14.1239332 Z"
              fill="#fff"
            />
          </g>
        </svg>
      </a>
      <nav
        class="nav"
      >
        <a
          class="link"
          href="/"
        >
          About
        </a>
        <a
          class="link"
          href="/portfolio"
        >
          Portfolio
        </a>
        <a
          class="link"
          href="/posts"
        >
          Ramblings
        </a>
        <a
          class="link"
          href="/uses"
        >
          Uses
        </a>
      </nav>
      <a
        href="https://www.github.com/visormatt"
        rel="noopener noreferrer"
        target="_blank"
        title="Find me on Github"
      >
        <svg
          class="svgOctocat"
          height="50"
          viewBox="0 0 250 250"
          width="50"
        >
          <path
            d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"
            fill="#fff"
          />
          <path
            class="octocat-arm"
            d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5, 78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5, 87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2"
            fill="#222"
          />
          <path
            class="octocat-body"
            d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3, 112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z"
            fill="#222"
          />
        </svg>
      </a>
    </div>
  </div>,
]
`;
